File: Top Interview 150/Hard/MergeKSortedLists.java
final class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

public class MergeKSortedLists {
    MergeKSortedLists () {}
    
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) return null;
        return mergeKLists(lists, 0, lists.length - 1);
    }

    private ListNode mergeKLists(ListNode[] lists, int start, int end) {
        if (start == end) return lists[start];
        int mid = (start + end) >> 1;
        ListNode l1 = mergeKLists(lists, start, mid);
        ListNode l2 = mergeKLists(lists, mid + 1, end);
        return mergeTwoLists(l1, l2);
    }
    
    private ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode head = new ListNode();
        ListNode curr = head;
        
        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                curr.next = l1;
                l1 = l1.next;
            }
            else {
                curr.next = l2;
                l2 = l2.next;
            }
            curr = curr.next;
        }
        
        curr.next = l1 != null ? l1 : l2;
        return head.next;
    }

    public void printList (ListNode head) {
        if (head == null) return;
        System.out.println(head.val);
        printList(head.next);
    }

    public static void main (String[] args) {
        MergeKSortedLists sol = new MergeKSortedLists();
    
        ListNode l1 = new ListNode(1);
        l1.next = new ListNode(4);
        l1.next.next = new ListNode(5);
        
        ListNode l2 = new ListNode(1);
        l2.next = new ListNode(3);
        l2.next.next = new ListNode(4);
        
        ListNode l3 = new ListNode(-1);
        l3.next = new ListNode(6);
        
        ListNode[] arr = {l1, l2, l3};
        long startTime = System.nanoTime();
        ListNode head = sol.mergeKLists(arr);
        long endTime = System.nanoTime();
        long elapsedTime = endTime - startTime;
        System.out.printf("Runtime: %.10fs%n", elapsedTime / 1_000_000_000.0);
        sol.printList(head);
    }
}

File: Top Interview 150/Hard/MergeKSortedLists.py
import time

class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        if lists:
            length = len(lists)
        
            if length > 2:
                half = length >> 1
                return self.mergeKLists([self.mergeKLists(lists[:half]), self.mergeKLists(lists[half:])])

            l1 = lists[0]
            l2 = lists[1] if length > 1 else None
            head = ListNode()
            curr = head
            
            while l1 and l2:
                if l1.val < l2.val:
                    curr.next = l1
                    l1 = l1.next
                    curr = curr.next
                else:
                    curr.next = l2
                    l2 = l2.next
                    curr = curr.next
            
            if l1:
                curr.next = l1
            elif l2:
                curr.next = l2
            
            return head.next
        


def printList(head):
    if head is None:
        return
    print(head.val)
    printList(head.next)


if __name__ == "__main__":
    sol = Solution()
    
    l1 = ListNode(1)
    l1.next = ListNode(4)
    l1.next.next = ListNode(5)
    
    l2 = ListNode(1)
    l2.next = ListNode(3)
    l2.next.next = ListNode(4)
    
    l3 = ListNode(-1)
    l3.next = ListNode(6)
    start_time = time.time()
    head = sol.mergeKLists([l1, l2, l3])
    end_time = time.time()
    elapsed_time = end_time - start_time
    print(f"Runtime: {round(elapsed_time, 10)}s")
    printList(head)
